### PR TITLE
Routerpubsubtested

### PR DESCRIPTION
**Pull request by shwethanidd**
Changes done:
1. PubSubService is part of router. Created within concrete router in main.py
2. Pubsub subsystem sends all the messages using send_vip() with subsystem='pubsub'. Router checks the subsystem frame and dispatches to PubsubService.
3. Pubsub subsystem waits on any incoming messages (publish or replies to list queries etc) and takes appropriate action.
4. I have modified protected topics implementation after discussing with Zach. He needs to review the changes as well.
5. Done few corrections for pytest failures

---
This change is [Reviewable](https://reviewable.io/reviews/volttron/volttron/1097)

### CODE REVIEW

File: volttron/platform/main.py [hunk 1]
```diff
@@ -368,44 +381,54 @@ def handle_subsystem(self, frames, user_id):
...
+#             try:
+#                 for entry in write_protect:
+#                     topics.add(entry['topic'], entry['capabilities'])
+#             except KeyError:
+#                 _log.exception('invalid format for protected topics '
+#                                'file {}'.format(self._protected_topics_file))
+#             else:
+#                 self.vip.pubsub.protected_topics = topics
+#                 _log.info('protected-topics file %s loaded',
+#                           self._protected_topics_file)
```
Comment (kmonson) [inline volttron/platform/main.py]: We can delete this.

File: volttron/platform/main.py [hunk 2]
```diff
@@ -588,9 +614,9 @@ def router(stop):
             ControlService(opts.aip, address=address, identity='control',
                            tracker=tracker, heartbeat_autostart=True,
                            enable_store=False, enable_channel=True),
-            PubSubService(protected_topics_file, address=address,
-                          identity='pubsub', heartbeat_autostart=True,
-                          enable_store=False),
+            # PubSubService(protected_topics_file, address=address,
+            #               identity='pubsub', heartbeat_autostart=True,
+            #               enable_store=False),
```
Comment (kmonson) [inline volttron/platform/main.py]: We should delete this too.

File: volttron/platform/main.py [hunk 3]
```diff
@@ -368,44 +381,54 @@ def handle_subsystem(self, frames, user_id):
...
-                    topics.add(entry['topic'], entry['capabilities'])
-            except KeyError:
-                _log.exception('invalid format for protected topics '
-                               'file {}'.format(self._protected_topics_file))
-            else:
-                self.vip.pubsub.protected_topics = topics
-                _log.info('protected-topics file %s loaded',
-                          self._protected_topics_file)
+        elif subsystem == b'pubsub':
+            if self._pubsub is not None:
```
Comment (kmonson) [inline volttron/platform/main.py]: Can this ever be None?

File: volttron/platform/vip/agent/subsystems/ping.py [hunk 4]
```diff
@@ -82,6 +82,8 @@ def ping(self, peer, *args):
         result = next(self._results)
         args = list(args)
         args.insert(0, b'ping')
+        if peer == b'pubsub':
+            peer = b''
```
Comment (kmonson) [inline volttron/platform/vip/agent/subsystems/ping.py]: Why is this needed?

File: volttron/platform/vip/agent/subsystems/pubsub.py [hunk 5]
```diff
@@ -101,154 +104,48 @@ def __init__(self, core, rpc_subsys, peerlist_subsys, owner):
...
-            rpc_subsys.export(self._peer_unsubscribe, 'pubsub.unsubscribe')
-            rpc_subsys.export(self._peer_list, 'pubsub.list')
-            rpc_subsys.export(self._peer_publish, 'pubsub.publish')
-            rpc_subsys.export(self._peer_push, 'pubsub.push')
+            # rpc_subsys.export(self._peer_sync, 'pubsub.sync')
+            # rpc_subsys.export(self._peer_subscribe, 'pubsub.subscribe')
+            # rpc_subsys.export(self._peer_unsubscribe, 'pubsub.unsubscribe')
+            # rpc_subsys.export(self._peer_list, 'pubsub.list')
+            # rpc_subsys.export(self._peer_publish, 'pubsub.publish')
+            # rpc_subsys.export(self._peer_push, 'pubsub.push')
```
Comment (kmonson) [inline volttron/platform/vip/agent/subsystems/pubsub.py]: We should remove these.

File: volttron/platform/vip/agent/subsystems/ping.py (hunk 4 above)
Comment (craig8) [inline volttron/platform/vip/agent/subsystems/ping.py]: So b'' is the vip_identity of the router.  Is pubsub not addressable using the vip identity pubsub?  I suppose not since it is part of the router.  We need to put pubsub up a level at self.vip.publish(topic, message, headers=None) rather than under the self.vip.pubsub.  Because now the pubsub is useless in the call.

File: volttron/platform/vip/agent/subsystems/pubsub.py [hunk 6]
```diff
@@ -101,154 +104,41 @@ def __init__(self, core, rpc_subsys, peerlist_subsys, owner):
...
             core.onconnected.connect(self._connected)
-            core.onviperror.connect(self._viperror)
             peerlist_subsys.onadd.connect(self._peer_add)
-            peerlist_subsys.ondrop.connect(self._peer_drop)
-
+            self.vip_socket = self.core().socket
             def subscribe(member):   # pylint: disable=redefined-outer-name
                 for peer, bus, prefix in annotations(
                         member, set, 'pubsub.subscriptions'):
                     # XXX: needs updated in light of onconnected signal
```
Comment (craig8) [inline volttron/platform/vip/agent/subsystems/pubsub.py]: What is # XXX is that like a TODO?

File: volttron/platform/vip/agent/subsystems/pubsub.py [hunk 7]
```diff
@@ -101,154 +104,41 @@ def __init__(self, core, rpc_subsys, peerlist_subsys, owner):
...
                     self.add_subscription(peer, prefix, member, bus)
+                    #_log.debug("PUBSUB SUBSYS subscribe with annotate. prefix {0}, identity: {1}".format(prefix, self.core().identity))
```
Comment (craig8) [inline volttron/platform/vip/agent/subsystems/pubsub.py]: Either uncomment debug messages or remove them.  I would be in favor of uncommented even if it's more traffic in the logs.

File: volttrontesting/platform/auth_test.py [hunk 8]
```diff
@@ -187,16 +188,19 @@ def pubsub_authorized(volttron_instance_encrypt, topic='foo', regex=None, peer='
 
 
 @pytest.mark.auth
```
Comment (craig8) [inline volttrontesting/platform/auth_test.py]: Why are these marked as xfail?

File: volttron/platform/vip/agent/subsystems/ping.py (hunk 4 above)
Comment (kmonson) [inline volttron/platform/vip/agent/subsystems/ping.py]: I think moving publish to self.vip.publish is not a good idea. It breaks backward comparability.

Comment (jhaack) [inline volttron/platform/vip/agent/subsystems/ping.py]: If at all possible, we need to preserve backwards compatibility on such a central piece of the  platform.

Comment (kmonson) [inline volttron/platform/vip/agent/subsystems/ping.py]: I think we can remove test_can_ping_pubsub as that test works on the assumption that pubsub is implemented as a service agent and this is no longer the case. In fact, any test that is working on that assumption should be removed. Then we can remove this check as it's really not needed.

Comment (kmonson) [inline volttron/platform/vip/agent/subsystems/ping.py]: That would be a general exception to the previous prohibition against modifying test for this.

Comment (craig8) [inline volttron/platform/vip/agent/subsystems/ping.py]: NOTE I did not want to move.  I wanted to have an alias that calls the normal pubsub...but without the extra arguments.

File: volttron/platform/vip/agent/subsystems/pubsub.py (hunk 7 above)
Comment (shwethanidd) [inline volttron/platform/vip/agent/subsystems/pubsub.py]: Ok, I'll uncomment the important ones and remove the rest.

File: volttrontesting/platform/auth_test.py (hunk 8 above)
Comment (shwethanidd) [inline volttrontesting/platform/auth_test.py]: All the test_pubsub_unauthorized* tests are marked xfail for now since auth is sending empty capabilities list. @zbeech is looking into the failures.
test_pubsub_authorized is wrongly marked as xfail. I'll remove that.

File: volttron/platform/vip/agent/subsystems/ping.py (hunk 4 above)
Comment (kmonson) [inline volttron/platform/vip/agent/subsystems/ping.py]: @craig8 In that case it's a good idea.

Background in case anyone doesn't know what he's referring to in regards to extra arguments:

The "peer" argument to publish, subscribe, etc. is obsolete now as this change removed the (completely unused) ability to for an agent to have it's own personal pubsub. It's now a no op.

File: volttron/platform/vip/agent/subsystems/pubsub.py [hunk 9]
```diff
@@ -101,154 +104,41 @@ def __init__(self, core, rpc_subsys, peerlist_subsys, owner):
...
-            rpc_subsys.export(self._peer_unsubscribe, 'pubsub.unsubscribe')
-            rpc_subsys.export(self._peer_list, 'pubsub.list')
-            rpc_subsys.export(self._peer_publish, 'pubsub.publish')
-            rpc_subsys.export(self._peer_push, 'pubsub.push')
+            self._processgreenlet = gevent.spawn(self._process_loop)
             core.onconnected.connect(self._connected)
-            core.onviperror.connect(self._viperror)
             peerlist_subsys.onadd.connect(self._peer_add)
-            peerlist_subsys.ondrop.connect(self._peer_drop)
-
```
Comment (kmonson) [inline volttron/platform/vip/agent/subsystems/pubsub.py]: We should remove all the stuff from the subsystem that you've disconnected here. 

All that should be left in the subsystem code is the client side of the pubsub. Everything left over from the old service agent needs to go.

File: volttron/platform/vip/agent/subsystems/pubsub.py [hunk 10]
```diff
@@ -101,154 +104,41 @@ def __init__(self, core, rpc_subsys, peerlist_subsys, owner):
         self._peer_subscriptions = {}
         self._my_subscriptions = {}
```
Comment (kmonson) [inline volttron/platform/vip/agent/subsystems/pubsub.py]: As there is only one possible peer now so self._my_subscriptions doesn't need to be a dictionary, I think. Also anywhere that references it should now handle the peer argument correctly by ignoring it.

self._peer_subscriptions can go away as it's no longer used.

Comment (kmonson) [inline volttron/platform/vip/agent/subsystems/pubsub.py]: As there is only one possible peer now so self._my_subscriptions doesn't need to be a dictionary, I think. Also anywhere that references it should now handle the peer argument correctly by ignoring it.

self._peer_subscriptions can go away as it's no longer used.

**Comment by craig8** [comment]
Please pull from develop and commit that to this branch as well.

**Comment by shwethanidd** [comment]
Merged with develop

File: volttron/platform/vip/pubsubwrapper.py [hunk 11]
```diff
@@ -80,41 +81,71 @@ def decode_peer(peer):
...
     def onsetup(self, sender, **kwargs):
         # pylint: disable=unused-argument
+        self.vip.rpc.export(self._peer_sync, 'pubsub.sync')
         self.vip.rpc.export(self._peer_publish, 'pubsub.publish')
         self.vip.rpc.export(self._peer_subscribe, 'pubsub.subscribe')
         self.vip.rpc.export(self._peer_unsubscribe, 'pubsub.unsubscribe')
         self.vip.rpc.export(self._peer_list, 'pubsub.list')
 
+    def _sync(self, peer, items):
+        _log.debug("Im here???????????")
```
Comment (kmonson) [inline volttron/platform/vip/pubsubwrapper.py]: Remove this debug message.

File: volttron/platform/vip/pubsubwrapper.py [hunk 12]
```diff
@@ -80,41 +81,71 @@ def decode_peer(peer):
...
+            _log.debug("PUBSUBWRAPPER subscribing again")
+            self.vip.pubsub.subscribe(peer, prefix, self._collector, bus=bus)
+
+    def _peer_sync(self, items):
+        peer = bytes(self.vip.rpc.context.vip_message.peer)
+        assert isinstance(items, dict)
+        self._sync(peer, items)
+
     def _peer_publish(self, topic, headers, message=None, bus=''):
+        _log.debug("PUBSUBWRAPPER::perr_publish")
```
Comment (kmonson) [inline volttron/platform/vip/pubsubwrapper.py]: Remove this debug message.

File: volttron/platform/vip/pubsubwrapper.py [hunk 13]
```diff
@@ -80,41 +81,71 @@ def decode_peer(peer):
...
-        subscriptions = self._peer_subscriptions[bus]
-        try:
-            subscribers = subscriptions[prefix]
-        except KeyError:
-            subscriptions[prefix] = subscribers = set()
-        subscribers.add(peer)
+        self._peer_subscriptions[bus][prefix].add(peer)
 
     def _peer_subscribe(self, prefix, bus=''):
+        _log.debug("PUBSUBWRAPPER::perr_subscribe")
```
Comment (kmonson) [inline volttron/platform/vip/pubsubwrapper.py]: Remove this debug message.

File: volttron/platform/vip/pubsubwrapper.py [hunk 14]
```diff
@@ -163,6 +194,7 @@ def _peer_list(self, prefix='', bus='', subscribed=True, reverse=False):
         return results
 
     def _peer_unsubscribe(self, prefix, bus=''):
+        _log.debug("PUBSUBWRAPPER::perr_unsubscribe")
```
Comment (kmonson) [inline volttron/platform/vip/pubsubwrapper.py]: Remove this debug message.

File: volttron/platform/vip/pubsubwrapper.py [hunk 15]
```diff
@@ -179,4 +211,4 @@ def _peer_unsubscribe(self, prefix, bus=''):
                 subscribers.discard(peer)
                 if not subscribers:
                     del subscriptions[prefix]
-        self.vip.pubsub.unsubscribe(peer, prefix, self._collector, bus=bus)
+                    self.vip.pubsub.unsubscribe(peer, prefix, self._collector, bus=bus)
```
Comment (kmonson) [inline volttron/platform/vip/pubsubwrapper.py]: I think we also need to do this up up on line 207 after 

del subscriptions[topic]

**Comment by shwethanidd** [comment]
Removing this pull request. Instead creating a new pull request from "routerfixeswithujson" branch.
